File: packages/python/hudy/cache.py
"""Smart caching with year-based TTL."""

import time
from datetime import date, datetime
from typing import Dict, List, Optional

from .types import CacheStats, Holiday


class CacheEntry:
    """Cache entry with expiration."""

    def __init__(self, data: List[Holiday], expires_at: float):
        self.data = data
        self.expires_at = expires_at


class SmartCache:
    """Smart cache with year-based TTL."""

    def __init__(self, custom_ttl: Optional[int] = None):
        self._store: Dict[str, CacheEntry] = {}
        self._hits = 0
        self._misses = 0
        self._custom_ttl = custom_ttl  # User-provided override in seconds

    def get(self, year: int) -> Optional[List[Holiday]]:
        """Get cached holidays for a year."""
        key = self._get_year_key(year)
        entry = self._store.get(key)

        if not entry:
            self._misses += 1
            return None

        # Check if expired
        if time.time() > entry.expires_at:
            del self._store[key]
            self._misses += 1
            return None

        self._hits += 1
        return entry.data

    def set(self, year: int, holidays: List[Holiday]) -> None:
        """Store holidays for a year with appropriate TTL."""
        key = self._get_year_key(year)
        ttl = self._get_ttl(year)
        expires_at = time.time() + ttl

        self._store[key] = CacheEntry(holidays, expires_at)

    def get_range(self, from_date: date, to_date: date) -> Optional[List[Holiday]]:
        """
        Get cached holidays for a date range.
        Returns None if any required year is not fully cached.
        Note: Client now fetches by year and filters, so this may not be used directly.
        """
        from_year = from_date.year
        to_year = to_date.year

        from_str = from_date.strftime("%Y-%m-%d")
        to_str = to_date.strftime("%Y-%m-%d")

        # If single year, just check that year's cache
        if from_year == to_year:
            cached = self.get(from_year)
            if not cached:
                return None

            # Filter by date range
            return [h for h in cached if from_str <= h.date <= to_str]

        # Multi-year range: need all years cached
        all_holidays: List[Holiday] = []
        for year in range(from_year, to_year + 1):
            cached = self.get(year)
            if not cached:
                # Don't increment misses again - get() already did it
                return None
            all_holidays.extend(cached)

        # Don't increment hits again - the get() calls already did it
        return [h for h in all_holidays if from_str <= h.date <= to_str]

    def clear(self) -> None:
        """Clear all cached data."""
        self._store.clear()
        self._hits = 0
        self._misses = 0

    def get_stats(self) -> CacheStats:
        """Get cache statistics."""
        return CacheStats(
            hits=self._hits,
            misses=self._misses,
            size=len(self._store),
            entries=[
                {"key": key, "expiresAt": entry.expires_at}
                for key, entry in self._store.items()
            ],
        )

    def prune(self) -> int:
        """Remove expired entries (garbage collection)."""
        now = time.time()
        to_remove = [
            key for key, entry in self._store.items() if now > entry.expires_at
        ]

        for key in to_remove:
            del self._store[key]

        return len(to_remove)

    # Private helpers

    def _get_year_key(self, year: int) -> str:
        """Get cache key for a year."""
        return f"year:{year}"

    def _get_ttl(self, year: int) -> int:
        """Get TTL in seconds based on year."""
        # If user provided custom TTL, use it
        if self._custom_ttl is not None:
            return self._custom_ttl

        # Otherwise use year-based logic
        current_year = datetime.now().year

        if year < current_year:
            # Past: 1 year (very stable)
            return 365 * 24 * 60 * 60
        elif year == current_year:
            # Current: 1 day (may change)
            return 24 * 60 * 60
        else:
            # Future: 1 week (not finalized)
            return 7 * 24 * 60 * 60

File: packages/typescript/src/cache.ts
/**
 * Smart caching implementation
 *
 * Implements year-based TTL strategy:
 * - Past years (< current): 1 year TTL (very stable)
 * - Current year: 1 day TTL (may change)
 * - Future years (> current): 1 week TTL (not finalized)
 */

import { Holiday } from './types';

interface CacheEntry<T> {
  data: T;
  expiresAt: number; // timestamp in ms
}

export interface CacheStats {
  hits: number;
  misses: number;
  size: number;
  entries: Array<{
    key: string;
    expiresAt: number;
  }>;
}

export class SmartCache {
  private store: Map<string, CacheEntry<Holiday[]>>;
  private hits: number = 0;
  private misses: number = 0;
  private customTTL?: number; // User-provided override in seconds

  constructor(options?: { ttl?: number }) {
    this.store = new Map();
    this.customTTL = options?.ttl;
  }

  /**
   * Get cached holidays for a year
   */
  get(year: number): Holiday[] | null {
    const key = this.getYearKey(year);
    const entry = this.store.get(key);

    if (!entry) {
      this.misses++;
      return null;
    }

    // Check if expired
    if (Date.now() > entry.expiresAt) {
      this.store.delete(key);
      this.misses++;
      return null;
    }

    this.hits++;
    return entry.data;
  }

  /**
   * Store holidays for a year with appropriate TTL
   */
  set(year: number, holidays: Holiday[]): void {
    const key = this.getYearKey(year);
    const ttl = this.getTTL(year);
    const expiresAt = Date.now() + ttl * 1000;

    this.store.set(key, {
      data: holidays,
      expiresAt,
    });
  }

  /**
   * Get cached holidays for a date range.
   * Returns null if any required year is not fully cached.
   * Note: Client now fetches by year and filters, so this may not be used directly.
   */
  getRange(from: Date, to: Date): Holiday[] | null {
    const fromYear = from.getFullYear();
    const toYear = to.getFullYear();

    const fromStr = this.formatDate(from);
    const toStr = this.formatDate(to);

    // If single year, just check that year's cache
    if (fromYear === toYear) {
      const cached = this.get(fromYear);
      if (!cached) return null;

      return cached.filter((h) => h.date >= fromStr && h.date <= toStr);
    }

    // Multi-year range: need all years cached
    const allHolidays: Holiday[] = [];
    for (let year = fromYear; year <= toYear; year++) {
      const cached = this.get(year);
      if (!cached) {
        // Don't increment misses again - get() already did it
        return null;
      }
      allHolidays.push(...cached);
    }

    // Don't increment hits again - the get() calls already did it
    return allHolidays.filter((h) => h.date >= fromStr && h.date <= toStr);
  }

  /**
   * Clear all cached data
   */
  clear(): void {
    this.store.clear();
    this.hits = 0;
    this.misses = 0;
  }

  /**
   * Get cache statistics
   */
  getStats(): CacheStats {
    return {
      hits: this.hits,
      misses: this.misses,
      size: this.store.size,
      entries: Array.from(this.store.entries()).map(([key, entry]) => ({
        key,
        expiresAt: entry.expiresAt,
      })),
    };
  }

  /**
   * Remove expired entries (garbage collection)
   */
  prune(): number {
    const now = Date.now();
    let removed = 0;

    for (const [key, entry] of this.store.entries()) {
      if (now > entry.expiresAt) {
        this.store.delete(key);
        removed++;
      }
    }

    return removed;
  }

  // Private helpers

  private getYearKey(year: number): string {
    return `year:${year}`;
  }

  /**
   * Get TTL in seconds based on year
   */
  private getTTL(year: number): number {
    // If user provided custom TTL, use it
    if (this.customTTL !== undefined) {
      return this.customTTL;
    }

    // Otherwise use year-based logic
    const currentYear = new Date().getFullYear();

    if (year < currentYear) {
      // Past: 1 year (very stable)
      return 365 * 24 * 60 * 60;
    } else if (year === currentYear) {
      // Current: 1 day (may change)
      return 24 * 60 * 60;
    } else {
      // Future: 1 week (not finalized)
      return 7 * 24 * 60 * 60;
    }
  }

  private formatDate(date: Date): string {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
  }
}
